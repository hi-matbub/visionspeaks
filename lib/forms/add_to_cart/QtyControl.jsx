import React, { useState } from 'react';
import PropTypes from 'prop-types';
import Form from 'react-bootstrap/Form';
import Alert from '../../components/Alert';

const QtyControl = ({
  product,
  cart,
  handleUpdatedCartInState,
  currentBill,
}) => {
  const [alert, setAlert] = useState(null);

  const handleSubmit = (ev) => {
    ev.preventDefault();

    let updatedCart;
    let billing;
    const due = product.metadata.price * ev.target.qty.value;

    if (cart) {
      const newBilling = currentBill + due;

      updatedCart = JSON.stringify({
        ...cart,
        [product.id]: {
          ...product,
          price: product.metadata.price,
          qty: ev.target.qty.value,
        },
      });

      billing = JSON.stringify(newBilling);
    } else {
      // if cart empty, set cart
      updatedCart = JSON.stringify({
        [product.id]: {
          ...product,
          price: product.metadata.price,
          qty: ev.target.qty.value,
        },
      });

      billing = JSON.stringify(due);
    }

    localStorage.setItem('cart', updatedCart);
    localStorage.setItem('billing', billing);

    handleUpdatedCartInState(updatedCart, billing);

    setAlert({
      status: 'success',
      message:
        ev.target.qty.value === 1
          ? `added ${ev.target.qty.value} product to your cart!`
          : `added ${ev.target.qty.value} products to your cart!`,
    });
    setTimeout(() => {
      setAlert(null);
    }, 3000);
  };

  return (
    <div className="d-flex justify-content-end">
      <div
        style={{
          height: '100px',
          position: 'absolute',
          zIndex: '2000',
          left: '1em',
          bottom: '-1em',
        }}
      >
        {alert ? <Alert status={alert.status}>{alert.message}</Alert> : null}
      </div>

      <div className="d-flex flex-column align-items-end mb-3">
        <Form className="d-flex align-items-center" onSubmit={handleSubmit}>
          <h5 className="p-0 m-0">QTY:</h5>
          <Form.Control
            type="number"
            label="qty"
            min="1"
            max="40"
            name="qty"
            defaultValue="1"
            className="border-dark mx-3"
            style={{
              width: 'fit-content',
            }}
          />
          <button
            type="submit"
            className="btn btn-outline-dark qty_submit"
            style={{ width: '125px !important' }}
          >
            Add To Cart
          </button>
        </Form>
      </div>
    </div>
  );
};

QtyControl.propTypes = {
  product: PropTypes.shape({
    id: PropTypes.string.isRequired,
    metadata: PropTypes.shape({
      price: PropTypes.string.isRequired,
    }).isRequired,
  }).isRequired,
  cart: PropTypes.shape({}),
  handleUpdatedCartInState: PropTypes.func.isRequired,
  currentBill: PropTypes.number,
};

QtyControl.defaultProps = {
  cart: {},
  currentBill: 0,
};

export default QtyControl;
