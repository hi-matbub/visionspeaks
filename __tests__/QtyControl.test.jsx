import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import QtyControl from '../lib/forms/add_to_cart/QtyControl';

const id = 'demo_sku';

const product = {
  name: 'product',
  description: 'this is an awesome product you should buy 2',
  price: 287,
  id: 'abc_123',
  images: [
    '/path_01.png',
    '/path_02.png',
    '/path_03.png',
  ],
  qty: 12,
};

const handleUpdatedCartInState = () => {};

it('QtyControl component renders smoke test without crashing', () => {
  const div = document.createElement('div');
  ReactDOM.render(
    <QtyControl 
      id={id}
      product={product}             
      handleUpdatedCartInState={handleUpdatedCartInState}    
    />
  , div);
  ReactDOM.unmountComponentAtNode(div);
});

it('QtyControl component renders unchanged', () => {
  const tree = renderer.create(
    <QtyControl 
      id={id} 
      product={product}             
      handleUpdatedCartInState={handleUpdatedCartInState}    
    />
    ).toJSON();
  expect(tree).toMatchSnapshot();
});
