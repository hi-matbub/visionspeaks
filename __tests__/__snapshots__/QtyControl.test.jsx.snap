// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QtyControl component renders unchanged 1`] = `
<div
  className="d-flex justify-content-end"
>
  <div
    style={
      Object {
        "bottom": "-1em",
        "height": "100px",
        "left": "1em",
        "position": "absolute",
        "zIndex": "2000",
      }
    }
  />
  <div
    className="d-flex flex-column align-items-end mb-3"
  >
    <form
      className="d-flex align-items-center"
      onSubmit={[Function]}
    >
      <h5
        className="p-0 m-0"
      >
        QTY:
      </h5>
      <input
        className="border-dark mx-3 form-control"
        defaultValue="1"
        label="qty"
        max="40"
        min="1"
        name="qty"
        style={
          Object {
            "width": "fit-content",
          }
        }
        type="number"
      />
      <button
        className="btn btn-outline-dark qty_submit"
        style={
          Object {
            "width": "125px !important",
          }
        }
        type="submit"
      >
        Add To Cart
      </button>
    </form>
  </div>
</div>
`;
